packages/core/src/errors.ts

```diff
@@ -1,58 +1,176 @@
+import { ERRORS, ErrorDescriptor, getErrorCode } from "./errors-list";
+
 /**
- * All exceptions intenionally thrown with Ignition-core
- * extend this class.
+ * Base error class extended by all custom errors.
+ * Placeholder to allow us to customize error output formatting in the future.
  *
  * @alpha
  */
-export class IgnitionError extends Error {
+export class CustomError extends Error {
   constructor(message: string) {
     super(message);
+    this.name = this.constructor.name;
+  }
+}
+
+/**
+ * All exceptions intentionally thrown with Ignition-core
+ * extend this class.
+ *
+ * @alpha
+ */
+export class IgnitionError extends CustomError {
+  constructor(
+    errorDescriptor: ErrorDescriptor,
+    messageArguments: Record<string, string | number> = {}
+  ) {
+    const prefix = `${getErrorCode(errorDescriptor)}: `;
+    const formattedMessage = applyErrorMessageTemplate(
+      errorDescriptor.message,
+      messageArguments
+    );
+
+    super(prefix + formattedMessage);
 
     this.name = this.constructor.name;
   }
 }
 
 /**
- * This error class represents issue detected by Ignition-cores
- * validation phase on the user inputed module. Validation errors
- * capture the stack to the action within the offending module,
- * to enhance the locality of the validation error message.
+ * This class is used to throw errors from Ignition plugins made by third parties.
  *
- * * @alpha
+ * @alpha
  */
-export class IgnitionValidationError extends IgnitionError {
-  constructor(message: string) {
+export class IgnitionPluginError extends CustomError {
+  public static isIgnitionPluginError(
+    error: any
+  ): error is IgnitionPluginError {
+    return (
+      typeof error === "object" &&
+      error !== null &&
+      (error as IgnitionPluginError)._isIgnitionPluginError === true
+    );
+  }
+
+  private readonly _isIgnitionPluginError = true;
+
+  public readonly pluginName: string;
+
+  constructor(pluginName: string, message: string) {
     super(message);
+    this.pluginName = pluginName;
+    Object.setPrototypeOf(this, IgnitionPluginError.prototype);
```

Comment:
Can we leave the comment that was here before? also, `new.target` is dynamic, so it works with extensions to this tupe.